src/vivarium_cluster_tools/psimulate/results/processing.py

```diff
@@ -52,76 +51,70 @@ def write_results_batch(
+    to_concat = to_concat_old + [df.reset_index(drop=True) for df in new_metadata]
 
     updated = _concat_preserve_types(to_concat)
 
     end = time.time()
-    logger.info(f"Concatenated {len(new)} metadata in {end - start:.2f}s.")
+    logger.info(f"Concatenated {len(new_metadata)} metadata in {end - start:.2f}s.")
     return updated
 
 
 def _concat_results(
-    old: Dict[str, pd.DataFrame], new: List[Dict[str, pd.DataFrame]]
+    old_results: Dict[str, pd.DataFrame], new_results: List[Dict[str, pd.DataFrame]]
 ) -> Dict[str, pd.DataFrame]:
     # Skips all the pandas index checking because columns are in the same order.
     start = time.time()
     results = {}
-    metrics = {key for new_results in new for key in new_results.keys()}
+    metrics = {key for new in new_results for key in new.keys()}
     for metric in metrics:
-        to_concat = [
+        to_concat_old = (
+            [old_results[metric].reset_index(drop=True)] if metric in old_results else []
+        )
+        to_concat = to_concat_old + [
             df.reset_index(drop=True)
-            for result in new
+            for result in new_results
             for met, df in result.items()
             if met == metric
         ]
 
-        if metric in old:
-            to_concat += [old[metric].reset_index(drop=True)]
-
         results[metric] = _concat_preserve_types(to_concat)
 
     end = time.time()
-    logger.info(f"Concatenated {len(new)} results in {end - start:.2f}s.")
+    logger.info(f"Concatenated {len(new_results)} results in {end - start:.2f}s.")
     return results
 
 
 def _concat_preserve_types(df_list: List[pd.DataFrame]) -> pd.DataFrame:
-    """Concatenation preserves all ``numpy`` dtypes but does not preserve any
-    pandas specific dtypes (e.g., categories become objects."""
-    # We assume that all dataframes in the list have identical dtypes to the first
-    # Also, we convert dtypes to their string representations to avoid comparison
-    # issues (especially with CategoricalDtype)
-    col_order = df_list[0].columns
-    dtypes = df_list[0].dtypes.astype(str)
-    columns_by_dtype = [list(dtype_group.index) for _, dtype_group in dtypes.groupby(dtypes)]
-
-    splits = []
-    for columns in columns_by_dtype:
-        original_dtypes = {col: df_list[0][col].dtype for col in columns}
-        slices = [df.filter(columns) for df in df_list]
-        slice_df = pd.DataFrame(data=np.concatenate(slices), columns=columns)
-        for col, dtype in original_dtypes.items():
-            slice_df[col] = slice_df[col].astype(dtype)
-        splits.append(slice_df)
-    return pd.concat(splits, axis=1)[col_order]
+    """Concatenate datasets and preserve all ``numpy`` dtypes (but not any
+    pandas-specific dtypes, e.g. categories become objects). We assume that all
+    dataframes in the list have identical columns and dtypes to the first.
+    """
+    dtype_mapping = {col: df_list[0][col].dtype for col in df_list[0].columns}
```

Comment:
What if df_list has no elements? Is this possible?

Reply:
Nope, it won't ever enter `write_results_batch` unless there are unwritten results and to to_concat will have to have something.